docs/enhancements/bmac-reboot-hosts-on-iso-change.md

```diff
@@ -0,0 +1,221 @@
+using obsolete artifacts will have:
+* a label indicating that it was booted from obsolete artifacts
+* an array in its Status of references to the resources that are obsolete
+
+### assisted-service InfraEnv controller
+
+The InfraEnv ISO URL changes when the ISO contents change. This makes it
+obvious to an API user that they have booted a host from an obsolete ISO,
+because the URLs won't match.
+
+When assisted-service creates an ISO, it will capture and include in the ISO:
+* the name, namespace, UID, and generation of the InfraEnv
+* the name, namespace, UID, and generation of each NMStateConfig
+
+At runtime, those resource identifiers for the InfraEnv and the NMStateConfig
+it used will be passed by the agent to assisted-service and added to the
+Agent's Status.
+
+### assisted-service Agent controller
+
+When the Agent resource gets created, its status should include the
+above-described reference details for:
+* The InfraEnv
+* The NMStateConfig it utilized during its boot
+
+Upon comparing those references to the resources that currently exist, if any
+of the referenced resources have changed, then the controller will add a label
+to the Agent resource indicating such. It will also add a reference to each
+obsolete resource to an array in the Agent's Status.
+
+### Bare Metal Agent Controller
+
+The Bare Metal Agent Controller will watch for Agents to have that label.
+
+If the Agent's NMStateConfig is obsolete, then it will reboot the corresponding
+BareMetalHost with the latest ISO URL.
+
+Else if the Agent's InfraEnv is obsolete, and the Agent does not have an associated
+NMStateConfig, then the BMAC will reboot the host only if finds a new
+NMStateConfig that matches the Agent.
+
+The BMAC will also watch for the InfraEnv URL to change, and if there are any
+BareMetalHosts that were booted with an obsolete URL but do not yet have an
+Agent resource, those BareMetalHosts will be rebooted with the latest ISO URL.
+
+ +---------------------------+
+ |                           |
+ |   Agent has the label?    |
+ |                           |
+ +----+-----------+----------+
+      |           |
+      v           v
+ +----+--+    +---+---+      +-------------+
+ |       |    |       |      |             |
+ |  Yes  |    |  No   +----->+  No action  |
+ |       |    |       |      |             |
+ +---+---+    +-------+      +-------------+
+     |
+     v
+ +---+----------------+
+ |                    |
+ |  Agent Status      |
+ |  cites an obsolete |
+ |  NMStateConfig?    |
+ |                    |
+ +---+------------+---+
+     |            |
+     v            v
+ +---+---+    +---+---+
+ |       |    |       |
+ |  Yes  |    |  No   |
+ |       |    |       |
+ +---+---+    +---+---+
+     |            |
+     |            v
+     |         +--+-------------------------+
+     |         |                            |
+     |         |  Is there a NMStateConfig  |
```

Comment:
why does nmstate need to search for match isn't it something that the agent will do?

Reply:
1. A host boots, registers, and an Agent resource gets created.
2. Time passes.
3. Someone creates a new NMStateConfig resource matching that host.

In that scenario, the Agent is not going to look for new NMStateConfig in the API. The only way to apply the new NMStateConfig will be to reboot that host with a new ISO.

Reply:
So it require host to match only creation, that mean that when someone delete nmstate then it will require reboot as well.
it's related to my comments from https://github.com/openshift/assisted-service/pull/2144#discussion_r662490993 
Maybe i misunderstand something

Reply:
Can you increase the verbosity of your concern? Perhaps describe a specific scenario step-by-step? I'm not sure what you're getting at.

Reply:
sorry you can resolve this comment, i read it again and now it's clear.